pinot-broker/src/main/java/org/apache/pinot/broker/requesthandler/BaseSingleStageBrokerRequestHandler.java

```diff
@@ -893,6 +897,24 @@ private CompileResult compileRequest(long requestId, String query, SqlNodeAndOpt
     AuthorizationResult authorizationResult =
         accessControl.authorize(httpHeaders, TargetType.TABLE, tableName, Actions.Table.QUERY);
 
+    //rewrite query
+    Map<String, String> queryOptions =
+        pinotQuery.getQueryOptions() == null ? new HashMap<>() : pinotQuery.getQueryOptions();
+    List<String> rowFilters =
+        authorizationResult.rowFilters().values().stream().flatMap(Collection::stream).collect(Collectors.toList());
+    StringBuilder sb = new StringBuilder();
+    for (int i = 0; i < rowFilters.size(); i++) {
+      sb.append(rowFilters.get(i));
+      if (i < rowFilters.size() - 1) {
+        sb.append(" AND ");
+      }
+    }
+    queryOptions.put("rowFilters", sb.toString());
+    pinotQuery.setQueryOptions(queryOptions);
+
+    CalciteSqlParser.queryRewrite(pinotQuery);
```

Comment:
Does the query have to rewritten here ? The rewriters are anyway called later when creating the broker request.

Reply:
At this point, we're actually in the broker request handler so the rewriters have been called before indeed. It's called [here](https://github.com/apache/pinot/blob/41f90f9ef81d6d91205df95c2685a3abbb0d4f89/pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotQueryResource.java#L285). `CalciteSqlParser#compileSqlNodeToPinotQuery` calls the query rewriters internally.

We'll have to figure out a better way to rewrite queries that depend on dynamic metadata being carried in query options (or some other place).